addons/crm/models/crm_lead.py

```diff
@@ -595,7 +595,16 @@ def write(self, vals):
         if any(field in ['active', 'stage_id'] for field in vals):
             self._handle_won_lost(vals)
 
-        write_result = super(Lead, self).write(vals)
+        # stage change between two won stages: does not chang the date_closed
+        if stage_is_won:
+            still_won_leads = self.filtered(lambda r: r.stage_id and r.stage_id.is_won)
+            new_won_leads = self - still_won_leads
+            still_won_vals = vals.copy()
+            del still_won_vals['date_closed']
```

Comment:
I don't really like "del" which is a strange statement in python. Can't we instead use "pop" which is a real function ? Like
```
leads_already_won = self.filtered(lambda r: r.stage_id.is_won)
if leads_already_won:
    result = super(Lead, leads_already_won).write(vals)
remaining = self - leads_already_won
if remaining:
    vals = vals.pop('date_closed', False)
    result = super(Leads, remaining).write(vals)
return result
```

Reply:
Doing like that if the first write() return value is False and the second is True then the global return value will be True while it should be false, isn't it? Doing something like that would solve the problem:
`
result = True
leads_already_won = self.filtered(lambda r: r.stage_id.is_won)
if leads_already_won:
    result = super(Lead, leads_already_won).write(vals)
remaining = self - leads_already_won
if remaining:
    vals.pop('date_closed', False)
    result = super(Leads, remaining).write(vals) and result
return result
`

Reply:
Write can't return False, it always returns True (see in models.py the base definition of write). We could even return directly True if we want, the returned result conveys no meaning. So no need to be that complicated with the return management of write :) .

Reply:
ok you reversed part of the logic by the way, the right operation is:
```
        leads_already_won = self.filtered(lambda r: r.stage_id.is_won)
        remaining = self - leads_already_won
        if remaining:
            result = super(Lead, remaining).write(vals)
        if leads_already_won:
            vals.pop('date_closed', False)
            result = super(Lead, leads_already_won).write(vals)
        return result
```